pkg/proc/core_linux_amd64.go

```diff
@@ -0,0 +1,248 @@
+package proc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+
+	"golang.org/x/debug/elf"
+	"golang.org/x/sys/unix"
+)
+
+const NT_FILE elf.NType = 0x46494c45 // "FILE".
+
+// readCore reads a core file from corePath corresponding to the executable at
+// exePath. For details on the Linux ELF core format, see:
+// http://www.gabriel.urdhr.fr/2015/05/29/core-file/,
+// http://uhlo.blogspot.fr/2012/05/brief-look-into-core-dumps.html,
+// elf_core_dump in http://lxr.free-electrons.com/source/fs/binfmt_elf.c,
+// and, if absolutely desperate, readelf.c from the binutils source.
+func readCore(corePath, exePath string) (*Core, error) {
+	core, err := elf.Open(corePath)
+	if err != nil {
+		return nil, err
+	}
+	exe, err := os.Open(exePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if core.Type != elf.ET_CORE {
+		return nil, fmt.Errorf("%v is not a core file", core)
+	}
+
+	notes, err := readNotes(core)
+	memory := buildMemory(core, exe, notes)
+
+	threads := map[int]*LinuxPrStatus{}
+	pid := 0
+	for _, note := range notes {
+		switch note.Type {
+		case elf.NT_PRSTATUS:
+			t := note.Desc.(*LinuxPrStatus)
+			threads[int(t.Pid)] = t
+		case elf.NT_PRPSINFO:
+			pid = int(note.Desc.(*LinuxPrPsInfo).Pid)
+		}
+	}
+	return &Core{
+		MemoryReader: memory,
+		Threads:      threads,
+		Pid:          pid,
+	}, nil
+}
+
+type Core struct {
+	MemoryReader
+	Threads map[int]*LinuxPrStatus
+	Pid     int
+}
+
+// Note is a note from the PT_NOTE prog.
+// Relevant types:
+// - NT_FILE: File mapping information, e.g. program text mappings. Desc is a LinuxNTFile.
+// - NT_PRPSINFO: Information about a process, including PID and signal. Desc is a LinuxPrPsInfo.
+// - NT_PRSTATUS: Information about a thread, including base registers, state, etc. Desc is a LinuxPrStatus.
+// - NT_FPREGSET (Not implemented): x87 floating point registers.
+// - NT_X86_XSTATE (Not implemented): Other registers, including AVX and such.
+type Note struct {
+	Type elf.NType
+	Name string
+	Desc interface{} // Decoded Desc from the
+}
+
+// readNotes reads all the notes from the notes prog in core.
+func readNotes(core *elf.File) ([]*Note, error) {
+	var notesProg *elf.Prog
+	for _, prog := range core.Progs {
+		if prog.Type == elf.PT_NOTE {
+			notesProg = prog
+			break
+		}
+	}
+
+	r := notesProg.Open()
+	notes := []*Note{}
+	for {
+		note, err := readNote(r)
+		if err != nil {
+			// Assume that any error is because we ran out of notes.
```

Comment:
I am confused - is it true that readNote returns an error only when we run out of notes? EOF? No possibility of corruption or bug in the api?